keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,982 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+We propose introducing new (additional, not replacing old one) API for requesting image pull,
+that will return stream with periodic updates sent through it to the client. The image pull
+request parameters will contain the type of update the client wants to receive aobut the progress:
+- time-based
+- size-based
+
+and the frequency of the updates:
+- every N seconds
+- every N Kibibytes of the total image size being downloaded
+respectively to the type
+
+If the client did not specify the preferred notification granularity, default values should be used,
+for instance every Gibibyte downloaded.
+
+This will be easy to use in CRI-compliant command-line-tools as well as kubelet to monitor the
+progress and publish events to the Pod object
+
+See design detail below.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The user is deploying an application to the k8s cluster, the Pod is being
+scheduled and creation of the Pod begins. The only event occurring on the Pod
+object events list is "Pulling image", with no progress indicator, or ETA.
+
+If CRI had a possility to expose the progress and / or ETA for completing the
+image pull, it would improve user experience greatly as well as debugging
+experience in situations when the image source is not available and the timeout
+will be inevitably reached, but not immediately: the progress events on the Pod
+object would indicate if the image transfer is ongoing at all.
+
+#### Story 2
+
+The user is debugging an issue in k8s cluster by running `crictl pull` command
+to pull the image - either to see if image is available or to check if connectivity
+to the image source is in place. If the image is available and the pulling
+is ongoing - no indication is possible to show based on CRI itself. At least
+containerd runtime is already providing such information based on the
+`ctr image pull docker.io/library/busybox:latest`.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+There should be a guidance on how the runtime should report the progress, for instance
+unpacking phase should be considered (see below).
+
+Pulling an image consists of two actions - downloading particular layer of the image
+and unpacking it. This repeats for every layer in the image.
+
+There is a technology that allows starting the container while it's not yet fully downloaded.
```

Comment:
For stargz snapshotter, the "image pull" will still happen from kubelet's point of view, but there will be no actual downloading because all the snapshots already "exist" [1]. I guess in this case, container runtime should just close the stream without sending anything back?



[1] https://github.com/containerd/stargz-snapshotter/blob/908d837b232e5850004fbfba1f3dc0e6710602c4/snapshot/snapshot.go#L272

Reply:
Yes, the same should happen when the whole image was pulled faster than the requested granularity period lapses.